syntax = "proto2";

message Model {
  extensions 1000 to max;
}

message FeatureConfig {
  // If true, use rgb feature.
  optional bool use_rgb = 1 [default = false];

  // If true, use flow feature.
  optional bool use_flow = 2 [default = false];

  // If true, use audio feature.
  optional bool use_audio = 3 [default = false];

  // If true, use gyroscope feature.
  optional bool use_gyroscope = 4 [default = false];

  // If true, use accelerator feature.
  optional bool use_accelerator = 5 [default = false];
}

message Conv1D {
  // Number of conv1d layers.
  optional uint32 layers = 1 [default = 1];

  // Dimensionality of the output space.
  optional uint32 filters = 2 [default = 100];

  // Length of the 1D convolution window.
  optional uint32 kernel_size = 3 [default = 3];

  // Dilation rate, value != 1 for atrou convolution.
  optional uint32 dilation_rate = 4 [default = 1];

  // Dropout rate.
  optional float dropout_rate = 5 [default = 0];
}

message FullyDet {
  extend Model {
    optional FullyDet ext = 1021;
  }

  // Number of noun classes.
  optional uint32 n_noun_classes = 2 [default = 300];

  // Number of verb classes.
  optional uint32 n_verb_classes = 3 [default = 97];

  // If specified, add a conv1d layer to filter the sequence.
  optional Conv1D conv1d = 4;

  // Path to the csv file storing video info.
  optional string video_info_csv_path = 5;

  // Path to the pkl file storing ground-truth.
  optional string eval_annotation_path = 6;

  // Feature config.
  optional FeatureConfig feature_config = 7;

  // Weight decay.
  optional float weight_decay = 10 [default = 0.0];

  // If true, train the audio model end-to-end.
  optional bool train_audio_model = 11 [default = false];

  optional string noun_classes_csv = 12 [default = 'epic-kitchens-100-annotations/EPIC_100_noun_classes.csv'];

  optional string verb_classes_csv = 13 [default = 'epic-kitchens-100-annotations/EPIC_100_verb_classes.csv'];
}

message WeaklyDet {
  extend Model {
    optional WeaklyDet ext = 1004;
  }

  // Number of noun classes.
  optional uint32 n_noun_classes = 2 [default = 300];

  // Number of verb classes.
  optional uint32 n_verb_classes = 3 [default = 97];

  // If specified, add a conv1d layer to filter the sequence.
  optional Conv1D conv1d = 4;

  // Path to the csv file storing video info.
  optional string video_info_csv_path = 5;

  // Path to the pkl file storing ground-truth.
  optional string eval_annotation_path = 6;

  // Weight of the attention sparsity loss.
  optional float sparsity_loss_weight = 7 [default = 0.0];

  // Dropout rate of the attention layer.
  optional float attention_dropout_rate = 9 [default = 0.0];

  // Feature config.
  optional FeatureConfig feature_config = 10;

  optional float weight_decay = 11 [default = 0.0];

  // If true, use gumbel softmax.
  optional bool use_gumbel_softmax = 8 [default = false];

  optional float gumbel_softmax_temperature = 12 [default = 0.5];

  optional float sparse_regularizer = 13 [default = 0.0];

  optional string noun_classes_csv = 14 [default = 'epic-kitchens-100-annotations/EPIC_100_noun_classes.csv'];

  optional string verb_classes_csv = 15 [default = 'epic-kitchens-100-annotations/EPIC_100_verb_classes.csv'];
}

message SingleTimeDet {
  extend Model {
    optional SingleTimeDet ext = 1005;
  }

  // Number of noun classes.
  optional uint32 n_noun_classes = 2 [default = 300];

  // Number of verb classes.
  optional uint32 n_verb_classes = 3 [default = 97];

  // If specified, add a conv1d layer to filter the sequence.
  optional Conv1D conv1d = 4;

  // Path to the csv file storing video info.
  optional string video_info_csv_path = 5;

  // Path to the pkl file storing ground-truth.
  optional string eval_annotation_path = 6;

  // Weight of the attention sparsity loss.
  optional float sparsity_loss_weight = 7 [default = 0.0];

  // Dropout rate of the attention layer.
  optional float attention_dropout_rate = 9 [default = 0.0];

  // Feature config.
  optional FeatureConfig feature_config = 10;

  optional float weight_decay = 11 [default = 0.0];

  optional string noun_classes_csv = 14 [default = 'epic-kitchens-100-annotations/EPIC_100_noun_classes.csv'];

  optional string verb_classes_csv = 15 [default = 'epic-kitchens-100-annotations/EPIC_100_verb_classes.csv'];
}

message SingleTimeClassAwareDet {
  extend Model {
    optional SingleTimeClassAwareDet ext = 1009;
  }

  // Number of noun classes.
  optional uint32 n_noun_classes = 2 [default = 300];

  // Number of verb classes.
  optional uint32 n_verb_classes = 3 [default = 97];

  // If specified, add a conv1d layer to filter the sequence.
  optional Conv1D conv1d = 4;

  // Path to the csv file storing video info.
  optional string video_info_csv_path = 5;

  // Path to the pkl file storing ground-truth.
  optional string eval_annotation_path = 6;

  // Weight of the attention sparsity loss.
  optional float sparsity_loss_weight = 7 [default = 0.0];

  // Dropout rate of the attention layer.
  optional float attention_dropout_rate = 9 [default = 0.0];

  // Feature config.
  optional FeatureConfig feature_config = 10;

  optional float weight_decay = 11 [default = 0.0];

  optional string noun_classes_csv = 14 [default = 'epic-kitchens-100-annotations/EPIC_100_noun_classes.csv'];

  optional string verb_classes_csv = 15 [default = 'epic-kitchens-100-annotations/EPIC_100_verb_classes.csv'];

  // If true, use gumbel softmax.
  optional bool use_gumbel_softmax = 21 [default = false];

  optional float gumbel_softmax_temperature = 22 [default = 0.5];

  optional float sparse_regularizer = 23 [default = 0.0];
}

message SingleTimeClassAgnosticDet {
  extend Model {
    optional SingleTimeClassAgnosticDet ext = 1019;
  }

  // Number of noun classes.
  optional uint32 n_noun_classes = 2 [default = 300];

  // Number of verb classes.
  optional uint32 n_verb_classes = 3 [default = 97];

  // If specified, add a conv1d layer to filter the sequence.
  optional Conv1D conv1d = 4;

  // Path to the csv file storing video info.
  optional string video_info_csv_path = 5;

  // Path to the pkl file storing ground-truth.
  optional string eval_annotation_path = 6;

  // Weight of the attention sparsity loss.
  optional float sparsity_loss_weight = 7 [default = 0.0];

  // Dropout rate of the attention layer.
  optional float attention_dropout_rate = 9 [default = 0.0];

  // Feature config.
  optional FeatureConfig feature_config = 10;

  optional float weight_decay = 11 [default = 0.0];

  optional string noun_classes_csv = 14 [default = 'epic-kitchens-100-annotations/EPIC_100_noun_classes.csv'];

  optional string verb_classes_csv = 15 [default = 'epic-kitchens-100-annotations/EPIC_100_verb_classes.csv'];

  // If true, use gumbel softmax.
  optional bool use_gumbel_softmax = 21 [default = false];

  optional float gumbel_softmax_temperature = 22 [default = 0.5];

  optional float sparse_regularizer = 23 [default = 0.0];
}

message SingleTimeClassAgnosticCrossAttentionDet {
  extend Model {
    optional SingleTimeClassAgnosticCrossAttentionDet ext = 1018;
  }

  // Number of noun classes.
  optional uint32 n_noun_classes = 2 [default = 300];

  // Number of verb classes.
  optional uint32 n_verb_classes = 3 [default = 97];

  // If specified, add a conv1d layer to filter the sequence.
  optional Conv1D conv1d = 4;

  // Path to the csv file storing video info.
  optional string video_info_csv_path = 5;

  // Path to the pkl file storing ground-truth.
  optional string eval_annotation_path = 6;

  // Weight of the attention sparsity loss.
  optional float sparsity_loss_weight = 7 [default = 0.0];

  // Dropout rate of the attention layer.
  optional float attention_dropout_rate = 9 [default = 0.0];

  // Feature config.
  optional FeatureConfig feature_config = 10;

  optional FeatureConfig query_config = 16;

  optional float weight_decay = 11 [default = 0.0];

  optional string noun_classes_csv = 14 [default = 'epic-kitchens-100-annotations/EPIC_100_noun_classes.csv'];

  optional string verb_classes_csv = 15 [default = 'epic-kitchens-100-annotations/EPIC_100_verb_classes.csv'];

  // If true, use gumbel softmax.
  optional bool use_gumbel_softmax = 21 [default = false];

  optional float gumbel_softmax_temperature = 22 [default = 0.5];

  optional float sparse_regularizer = 23 [default = 0.0];
}

message SingleTimeClassAwareDetV2 {
  extend Model {
    optional SingleTimeClassAwareDetV2 ext = 1029;
  }

  // Number of noun classes.
  optional uint32 n_noun_classes = 2 [default = 300];

  // Number of verb classes.
  optional uint32 n_verb_classes = 3 [default = 97];

  // If specified, add a conv1d layer to filter the sequence.
  optional Conv1D conv1d = 4;

  // Path to the csv file storing video info.
  optional string video_info_csv_path = 5;

  // Path to the pkl file storing ground-truth.
  optional string eval_annotation_path = 6;

  // Weight of the attention sparsity loss.
  optional float sparsity_loss_weight = 7 [default = 0.0];

  // Dropout rate of the attention layer.
  optional float attention_dropout_rate = 9 [default = 0.0];

  // Feature config.
  optional FeatureConfig feature_config = 10;

  optional float weight_decay = 11 [default = 0.0];

  optional string noun_classes_csv = 14 [default = 'epic-kitchens-100-annotations/EPIC_100_noun_classes.csv'];

  optional string verb_classes_csv = 15 [default = 'epic-kitchens-100-annotations/EPIC_100_verb_classes.csv'];

  // If true, use gumbel softmax.
  optional bool use_gumbel_softmax = 21 [default = false];

  optional float gumbel_softmax_temperature = 22 [default = 0.5];

  optional float sparse_regularizer = 23 [default = 0.0];
}
